navigation/dead_reckoning/sam_dead_reckoning/sam_dead_reckoning/dr_odom_2_lat_lon.py

```diff
@@ -0,0 +1,284 @@
+                                                            return_debug_tuple=True)
+
+            if not transform_status:
+                self._log("say something here?")
+
+            return transform_status
+
+        except (LookupException, ConnectivityException, ExtrapolationException) as e:
+            self.get_logger().warn(f'Error while checking transform: {e}')
+            return False
+
+    def odom_callback(self, odom_msg):
+        """
+        Callback function for converting odometry messages to lat lon coordinates
+        """
+        # Check that the service is available
+        if not self.latlon_client.service_is_ready():
+            return
+
+        # Regulate the rate of status and latlon updates
+        # Might be better to move to a timer at some point...
+        current_odom_time = rcl_time_to_secs(self.get_clock().now())
+        if self.check_status_valid(current_odom_time):
+            self._log("Odom Callback")
+        else:
+            return
+
+        # Check that both the utm zone and the transform between odom and utm are valid
+        odom_frame_id = odom_msg.header.frame_id
+        self.determine_utm_zone()
+        transform_status = self.determine_transform_status(odom_frame_id)
+
+        if False in [self.utm_zone_status, transform_status]:
+            self._log(f"UTM status: {self.utm_zone_status} - Transform status: {transform_status}")
+            return
+
+        # Perform the odom -> utm conversion
+        try:
+            utm_transform = self.tf_buffer.lookup_transform(target_frame=self.utm_frame,
+                                                            source_frame=odom_msg.header.frame_id,
+                                                            time=rclpy.time.Time())
+
+            odom_utm_pose = self.transformed_odometry_to_pose(odom_msg, utm_transform, self.utm_frame)
+
+        except (LookupException, ConnectivityException, ExtrapolationException) as e:
+            self.get_logger().error(f'Failed to transform odometry: {e}')
+            return
+
+        # Construct real request
+        req = UTMLatLon.Request()
+        req.utm_points = []
+        req.lat_lon_points = []
+
+        ps = PointStamped()
+        ps.header.frame_id = f"utm_{self.utm_zone[0]}_{self.utm_zone[1]}"
+        ps.point.x = odom_utm_pose.position.x
+        ps.point.y = odom_utm_pose.position.y
+        req.utm_points.append(ps)
+
+        self.future = self.latlon_client.call_async(req)
+        self.future.add_done_callback(callback=self.lat_lon_service_response_callback)
+
+        # Note:
+        # The resulting GeoPoint from the lat lon service is published in the service response callback
+
+    def lat_lon_service_response_callback(self, future):
+        """
+        Callback function for handling th response
+
+        A response callback was used because the lat lon conversion service request was causing some sort of deadlocking
```

Comment:
Threading is hard :)
You could use a multi-threaded executor for the node to spin in, which _i think_ would help with the deadlock, but this works just as well. David's action server is a good example of such if needed.

Alternatively, you could use the latlon converter as a pure python object. The BT has an example of doing something like this with the dubins planner service.